crates/valence/src/inventory/validate.rs

```diff
@@ -0,0 +1,643 @@
+use valence_protocol::packet::c2s::play::click_slot::ClickMode;
+use valence_protocol::packet::c2s::play::ClickSlotC2s;
+
+use super::{Inventory, InventoryWindow};
+use crate::prelude::CursorItem;
+
+/// Validates a click slot packet enforcing that all fields are valid.
+pub(crate) fn validate_click_slot_impossible(
+    packet: &ClickSlotC2s,
+    player_inventory: &Inventory,
+    open_inventory: Option<&Inventory>,
+) -> bool {
+    if (packet.window_id == 0) != open_inventory.is_none() {
+        return false;
+    }
+
+    let max_slot = match open_inventory {
+        Some(inv) => inv.slot_count() + 36,
+        None => player_inventory.slot_count(),
+    };
+
+    // check all slot ids and item counts are valid
+    if !packet.slots.iter().all(|s| {
+        (0..=max_slot).contains(&(s.idx as u16))
+            && (0..=64).contains(&(s.item.as_ref().map(|i| i.count()).unwrap_or(0)))
```

Comment:
I see two problems here:
1. This assumes all items have a max stack size of 64.
2. This can't handle the situation where a stack with a weird stack count _already exists_ in the inventory. (Try giving yourself a stack with 100 items using `.set_count(100)`. Vanilla client seems to handle this just fine.)